src/client/testing/testController/common/server.ts

```
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import * as http from 'http';
+import { v4 as uuidv4 } from 'uuid';
+import { Disposable, Event, EventEmitter } from 'vscode';
+import {
+    ExecutionFactoryCreateWithEnvironmentOptions,
+    IPythonExecutionFactory,
+    SpawnOptions,
+} from '../../../common/process/types';
+import { DataReceivedEvent, ITestServer, TestCommandOptions } from './types';
+
+export class PythonTestServer implements ITestServer, Disposable {
+    private _onDataReceived: EventEmitter<DataReceivedEvent> = new EventEmitter<DataReceivedEvent>();
+
+    private uuids: Map<string, string>;
+
+    private server: http.Server;
+
+    constructor(private executionFactory: IPythonExecutionFactory, private readonly port: number) {
+        this.uuids = new Map();
+
+        const requestListener: http.RequestListener = async (request, response) => {
+            const buffers = [];
+
+            for await (const chunk of request) {
+                buffers.push(chunk);
+            }
+
+            const data = Buffer.concat(buffers).toString();
+
+            response.end();
+
+            const { uuid } = JSON.parse(data);
```

What happens on parse error here?

---

Test discovery silently fails. How would you surface the error?

---

`traceDebug` if there is a failure so we can try and replicate it?

---

Do you think we should surface it as an error node in the test explorer as well?

---

Thinking of it, surfacing it as an error node might be a bit dicey, since the server doesn't know which workspace this faulty test data belongs to. Will display it as a debug log.

---

for these ones `traceDebug` should be enough